### PR TITLE
Trade Functionality

### PR DESCRIPTION
**Pull request by AnimeshKC**
Trading of pieces with the Shift key now works, along with ensuring that trades only happen if the piece wasn't traded. 

Cancelling of trades has now been implemented.

### CODE REVIEW

File: client/src/GameState.ts [hunk 1]
```diff
@@ -60,5 +63,29 @@ export class GameState {
...
+            this.tradeState = TradeState.NoTrade;
+            this.tradingPlayerId = null;
+            this.socket.emit("clearTrade")
+        })
+        
     }
+            public emitTrade() {
+            this.socket.emit("playerTrade", this.playerId,
+                this.currentTetromino.type, 
+                this.tradeState)
```
Comment (AdityaHarvi) [inline client/src/GameState.ts:87]: missing semi-colons ;-;
(and on above portion as well)

File: server/index.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Import the express in typescript file
 import express from "express";
 import http from "http";
-import { Server } from "socket.io";
+import { Server, Socket } from "socket.io";
```
Comment (AdityaHarvi) [inline server/index.ts:4]: Importing `Socket` can be removed once the `Trade` class is moved into its own file.
You'll also want to make its own socket event format in `common/messages` so that it doesn't have access to *every* single event.

File: server/index.ts [hunk 3]
```diff
@@ -98,6 +99,31 @@ export function broadcastRemainingPlayers(playersNeeded: number) {
 //   console.log("Sending clients to Game Over Screen!")
 //   scoreboard.displaySceneGameOver();
 // }, 30000);
+class Trade {
+    //currentOffer should be a socket
+    currentOfferer: Socket<ClientToServerEvents, ServerToClientEvents, InterServerEvents, SocketData> | null = null;
```
Comment (AdityaHarvi) [inline server/index.ts:104]: This socket will need to be redefined when moving it to a different file

**Review by AdityaHarvi — CHANGES_REQUESTED** [review]

File: client/src/scene/SceneGameArena.ts [hunk 4]
```diff
@@ -206,8 +239,19 @@ export class SceneGameArena extends Phaser.Scene {
                 scene.gameState.currentTetromino.reportState()
             );
         }
+        if (tradeChanged) {
+            scene.gameState.emitTrade();
+        }
+    }
+    private updateFromTradeState(scene: SceneGameArena) {
```
Comment (AdityaHarvi) [inline client/src/scene/SceneGameArena.ts:246]: Don't think you need `scene` here. You can access `this` directly.

File: server/package.json [hunk 5]
```diff
@@ -17,6 +17,7 @@
 		"dotenv": "^14.3.2",
 		"express": "^4.17.2",
 		"path": "^0.12.7",
+		"prettier": "^2.5.1",
```
Comment (AdityaHarvi) [inline server/package.json:20]: This is conflicting with the linter we are already using. Should be removed.

File: client/src/scene/SceneGameArena.ts [hunk 6]
```diff
@@ -130,12 +135,17 @@ export class SceneGameArena extends Phaser.Scene {
...
+            this.trade = new TradeUI(this);
+        }
+
         // 12 fps
         if (this.frameTimeElapsed > 1000 / this.FRAMERATE) {
             this.updateBoardFromFrozen(this, this.gameState.otherTetrominoes);
             this.updateUserInput(this);
             this.updateDrawBoard(this.gameState, this);
             this.updateDrawPlayer(this);
+            this.updateFromTradeState(this);
```
Comment (AdityaHarvi) [inline client/src/scene/SceneGameArena.ts:148]: What was the motivation for calling this function 12 times a second rather than just when the client receives a trade event?

File: server/tsconfig.json [hunk 7]
```diff
@@ -7,5 +7,5 @@
     "strict": true,
     "module": "CommonJS"
   },
-  "include": ["./src", "./index.ts"]
+  "include": ["./src", "./index.ts", "../common/PlayerAttributes.ts"]
```
Comment (AdityaHarvi) [inline server/tsconfig.json:10]: Why is this needed?

**Review by AdityaHarvi — CHANGES_REQUESTED** [review]

File: client/src/scene/ControlsUI.ts [hunk 8]
```diff
@@ -1,28 +1,29 @@
...
+  constructor(scene: SceneGameArena, keys: Array<string>) {
+    let y = 550;
+    let index = 0;
+    const controlInfo = [
+      "Move Left",
+      "Move Right",
+      "Move Down",
+      "Rotate CCW",
+      "Rotate CW",
+      "Trade Action"
```
Comment (AdityaHarvi) [inline client/src/scene/ControlsUI.ts:13]: Trade action is included into this file but the SVG is not being passed in? What is the purpose of keeping this here?

File: client/src/scene/SceneGameArena.ts [hunk 9]
```diff
@@ -130,12 +135,17 @@ export class SceneGameArena extends Phaser.Scene {
     update(time: number, delta: number) {
         this.frameTimeElapsed += delta;
 
+        if (this.trade == null && this.gameState.playerId != null) {
```
[the remark this answers is not in the document]
Comment (c25vdw) [reply client/src/scene/SceneGameArena.ts:137]: I think yes, the trade portion needs to be moved up. The reason I moved the controlsUI up is because it needed a `playerId` to identify players from spectators. Now that we are sending `initPlayer` to scene waiting room (meaning that scene game arena will have the playerID ready), there is no need to initialize components in update.

Speaking of `playerID` for player identification, I don't see that the TradeUI is in need of a playerID, which sounds like every one including spectator can see it right now. Correct me if I missed anything, but if that's the case maybe we should fix that in this PR (or another one).  

cc @AnimeshKC

**Comment by AdityaHarvi** [comment]
Since Eric is working on new artwork for the keys, you can safely delete the `key_SHIFT.svg` file

File: client/src/scene/SceneGameArena.ts [hunk 10]
```diff
@@ -177,6 +189,28 @@ export class SceneGameArena extends Phaser.Scene {
                 this.gameState.board,
                 Tetromino.rotateCW
             );
+        } else if (this.keys.shift.isDown) {
```
Comment (AdityaHarvi) [inline client/src/scene/SceneGameArena.ts:192]: Can you add in a note at the top of this function with like `FIXME: Spectators are able to press these buttons`

File: client/src/scene/SceneGameArena.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (AdityaHarvi) [reply client/src/scene/SceneGameArena.ts:137]: @c25vdw if we do remove the playerID check, then everyone will be able to view the currently ongoing trades. That is fine but it would mean we would have to hide the "shift key" portion of the UI so that spectators are not presented with that information.
So in either case, we will need to perform the playerID check

File: client/src/scene/SceneGameArena.ts (hunk 6 above)
Comment (AnimeshKC) [reply client/src/scene/SceneGameArena.ts]: Well, the client receives the trade event in gamestate, which doesn't have a direct way to update the scene text in SceneGameArena. So currently, Scene Game Arena updates based on the change of state in gamestate. 

This naturally causes a slight delay in the display which might occasionally be a problem.

File: client/src/GameState.ts [hunk 11]
```diff
@@ -95,6 +106,32 @@ export class GameState {
...
+            if (otherTradeState == TradeState.Offered) {
+                this.tradeState = TradeState.Pending;
+                this.tradingPlayerId = playerId;
+            } else if (otherTradeState == TradeState.Accepted) {
+                this.tradeState = TradeState.NoTrade;
+                this.tradingPlayerId = null;
+            }
+        });
+        this.socket.on("sendTradePiece", (tetrominoType) => {
+            this.currentTetromino.swapPiece(tetrominoType);
```
Comment (swebra) [inline client/src/GameState.ts:120]: I'm thinking that the `swapPiece` method should be removed all together and this should be replaced with just a re-instantiation of the current tetromino. Not only is `swapPiece` duplicate code from the Tetromino constructor, but it has functional issues due to pre-existing state that isn't present when constructing. For example, in the case of trading with the same type of tetromino, the `setType` would return early and thus the monomino positions and rotations would be out of sync from the Tetromino position and rotation.

We would need to destroy the existing monomino sprites before reassigning `this.tetromino` though, so Tetromino should get a `destroy()` method:
```ts
destroy() {
        if (this.monominoes) {
            this.monominoes.forEach((monomino) => monomino.destroy());
        }
}
```
and `Tetromino.setType` should be updated to call it. Then here it would be something like
```ts
        this.socket.on("sendTradePiece", (tetrominoType) => {
            this.currentTetromino.destroy();
            this.currentTetromino = new Tetromino(tetrominoType, this.playerId);
            // ...
```

File: client/src/GameState.ts [hunk 12]
```diff
@@ -95,6 +106,32 @@ export class GameState {
...
+            this.currentTetromino.isTraded = true;
+            this.clearTradeAndEmit();
```
Comment (swebra) [inline client/src/GameState.ts:122]: Should we also be calling `this.emitPlayerMove();` to update everyone with the new position?

File: client/src/GameState.ts [hunk 13]
```diff
@@ -121,6 +158,7 @@ export class GameState {
             this.currentTetromino.reportState()
         );
         this.placeTetromino(this.currentTetromino);
+        this.clearTradeAndEmit();
```
Comment (swebra) [inline client/src/GameState.ts:161]: Will this clear the trades for everyone? i.e. If player 1 puts their piece up to trade but player 2 places, won't that erroneously cancel player 1's trade?

File: client/src/scene/TradeUI.ts [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        [TradeState.Accepted]: "",
+    };
+    constructor(scene: SceneGameArena) {
+        const y = 50;
+        this.tradeState = TradeState.NoTrade;
+        this.tradingUser = null;
+        this.tradeText = scene.add.text(20, y, this.tradeMap[this.tradeState], {
+            fontSize: `20px`,
+            fontFamily: "VT323",
+        });
```
Comment (swebra) [inline client/src/scene/TradeUI.ts:24]: Looks like this has yet to be updated to use BitmapText? The trademap will accordingly need to be updated to be lowercase as well.

**Review by swebra — CHANGES_REQUESTED** [review]
This review is somewhat semi-complete as I haven't yet looked over all the exact trading logic + communications, but just wanted to get what I had thus far over to you. Most of this is just cleanup with the occasional functionality question. As a general comment, it would also be nice if the member function declarations could follow the existing style of having one newline between declarations.

File: client/src/GameState.ts (hunk 12 above)
Comment (AnimeshKC) [reply client/src/GameState.ts:198]: I guess this is probably necessary since position has changed. It's been added.

File: client/src/scene/SceneGameArena.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (AnimeshKC) [reply client/src/scene/SceneGameArena.ts:122]: Trade UI view for spectator will be handled in a different PR after the spectator view is resolved. Right now, enabling the Trade UI for non-players has strange effects on the spectator view.

File: client/src/scene/SceneGameArena.ts [hunk 15]
```diff
@@ -184,11 +197,47 @@ export class SceneGameArena extends Phaser.Scene {
             moved = this.gameState.moveIfCan(
                 Tetromino.rotateCW // clock wise
             );
+        } else if (this.keys.shift.isDown) {
```
Comment (AdityaHarvi) [inline client/src/scene/SceneGameArena.ts:200]: similar to the other key press handlers, this needs to check and ignore all spectator input (`this.gameState.playerId != null`)

**Review by AdityaHarvi — COMMENTED** [review]
LGTM, I'll just leave it to Eric for a last check over.

There are also some merge conflicts

File: client/src/scene/SceneGameArena.ts [hunk 16]
```diff
@@ -184,11 +197,47 @@ export class SceneGameArena extends Phaser.Scene {
             moved = this.gameState.moveIfCan(
                 Tetromino.rotateCW // clock wise
             );
+        } else if (this.keys.shift.isDown && this.gameState.playerId != null) {
```
Comment (swebra) [inline client/src/scene/SceneGameArena.ts:200]: I would maybe add the `!this.gameState.isInOppositeSection()` here too, though that's an assumption about the game behavior I'm making

**Review by swebra — CHANGES_REQUESTED** [review]
Just a few more notes and I'll reiterate the "newline between functions/methods" style again.

Also looks like I might change a bit of stuff around in TradeUI.ts when I prettify it to match everything else, but I wouldn't worry about those changes for this commit.

File: client/src/scene/SceneGameArena.ts (hunk 16 above)
Comment (AnimeshKC) [reply client/src/scene/SceneGameArena.ts:219]: Yeah not sure whether that's what we want. Otherwise, would already requested trades no longer be valid in opposite sections as well?

File: client/src/scene/SceneGameArena.ts [hunk 17]
```diff
@@ -74,7 +74,9 @@ export class SceneGameArena extends Phaser.Scene {
 
         this.scoreboard = new ScoreboardUI(this, this.socket);
         new ActiveEventsUI(this, this.socket);
-
+        if (this.gameState.playerId != null) {
+            this.trade = new TradeUI(this);
+        }
         if (this.gameState.playerId && this.gameState.playerId >= 0) {
             new ControlsUI(this);
         } else {
```
Comment (swebra) [inline client/src/scene/SceneGameArena.ts:82]: This doesn't need to be in its own if statement, can be just be in the if statement below

**Review by swebra — CHANGES_REQUESTED** [review]

File: client/src/scene/SceneGameArena.ts (hunk 17 above)
Comment (AdityaHarvi) [reply client/src/scene/SceneGameArena.ts:82]: Actually, that check might be broken. It would return false for the first check if playerId=0.

Can you replace `this.gameState.playerId && this.gameState.playerId >=0` with `this.gameState.playerId != null` and move your trade variable into it

Comment (AnimeshKC) [reply client/src/scene/SceneGameArena.ts]: Yeah that check is broken. Just changed it, but looks like the Controls UI was still working fine  because it is called when "initPlayer" is received.

**Review by swebra — APPROVED** [review]
